project/ArtiQule.py

```diff
@@ -0,0 +1,530 @@
+                                      "Design/icons/Sunbathing Eraser",
+                                      "Ctrl+F",
+                                      "Erase Your Mistakes, Kid!",
+                                      (99999, 99999))  # infinte duration
+
+        self.show()
+
+        self.drawing = False
+        self.lastPoint = QPoint()
+
+    def connectTool(self, curTool=None, curisDipped=False):
+
+        self.currentTool = curTool
+
+        try:
+            self.currentTool.duration
+            self.currentTool.isDipped
+        except AttributeError:
+            pass
+        else:
+            self.currentTool.isDipped = curisDipped
+            if self.currentTool.toolName == "Pointy Pen":
+                self.currentTool.duration = randint(
+                    0,
+                    self.currentTool.constDuration
+                )
+        ColorBox.setWindowCursor(self.currentTool)
+
+        if self.currentTool is not None:
+            self.setCursor(QCursor(
+                QPixmap("Design/icons/{}.png".format(self.currentTool.toolName)
+                        )
+                    )
+                )
+
+            if self.currentTool.toolName not in [
+                    "Pointy Pen",
+                    "A bucket",
+                    "Sunbathing Eraser",
+                    "A bucket filled"]:
+                if hasattr(self, "dripper"):
+                    self.dripper.stop()
+                self.dripper = DripperEffect(
+                    self,
+                    self.currentTool.brushSize
+                )
+                self.dripper.drip.connect(self.dripperHandler)
+                self.dripper.start()
+
+    def colorBoxRun(self):
+
+        self.colorBox = ColorBox(self)
+        ColorBox.setWindowCursor(self.currentTool)
+        geo = self.geometry()
+        geo.moveLeft(geo.right())  # moves window right
+        self.colorBox.setGeometry(geo)
+
+        number_of_buttons = 18
+        PalletteButtons = {PalletteButton(button, self): button for button in[
+                QPushButton() for _ in range(number_of_buttons)
+            ]
+        }
+
+        style_sheet_str = "background-color: rgba{0}; border-radius:20px"
+        for button_index in range(number_of_buttons):
+            c = list(PalletteButtons.keys())[button_index]
+            p = PalletteButtons[list(PalletteButtons.keys())[button_index]]
+            p.setStyleSheet(style_sheet_str.format(c.palletteColor))
+            # set the default value of c to c because
+            # otherwise it will always look up the newest value
+            p.clicked.connect(lambda state, c=c: c.mixColor(self.currentTool))
+            self.colorBox.addPallette(p)
+
+        # showing toolBox
+        self.colorBox.showColorBox()
+
+    def fill_bucket(self):
+        self.setCursor(QCursor(
+                QPixmap("Design/icons/{}.png".format("A bucket filled")
+                        )
```

Comment:
ahhhh this is the worst place this brace could be

Reply:
what do you mean by brace?

Reply:
brace as in parenthesis. If you were going to split this line up over multiple lines, you should be using something called K&R indentation style, or one of the other acceptable indentation styles. These braces are all over the place and this makes readability quite poor. 

K&R style indentation would look like this:
```py
    def fill_bucket(self):
        self.setCursor(
            QCursor(
                QPixmap("Design/icons/{}.png".format("A bucket filled"))
            )
        )
```

But to be honest, even if this was a single line, it would only be 88 characters, and would be within the 100 char limit we set in the `.flake8` file. so that would be acceptable, too. A little dense, though.

```py
    def fill_bucket(self):
        self.setCursor(QCursor(QPixmap("Design/icons/{}.png".format("A bucket filled"))))
```

I have no idea why you would .format a string literal into a string either, so we can shorten it by doing this:

```py
    def fill_bucket(self):
        self.setCursor(QCursor(QPixmap("Design/icons/A bucket filled.png")))
```

At this point readability is okay, but it's a bit unusual to have a filename with spaces, so let's change that too, while we're at it.

```py
    def fill_bucket(self):
        self.setCursor(QCursor(QPixmap("Design/icons/bucket_filled.png")))
```

But the problem with this is that it will fail on any platform that doesn't support the `/` for path selection. The solution to this is to involve pathlib.

```py
from pathlib import Path

ICONS = Path("Design", "icons")

def fill_bucket(self):
    # create a platform-agnostic path to the image
    # it's possible QPixmap doesn't support Path, so we can 
    # cast this to a string to make sure we pass it the same way
    filled_image = str(ICONS / "bucket_filled.png")

    self.setCursor(QCursor(QPixmap(filled_image)))
```

_now_ this looks pretty good.

Reply:
Good that I am now aware of that. Were there any other serious errors we could have avoided?